src/System.Windows.Forms/src/System/Windows/Forms/BindableComponent.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.ComponentModel;
+
+namespace System.Windows.Forms
+{
+    public abstract class BindableComponent : Component, IBindableComponent
+    {
+        internal static readonly object s_bindingContextChangedEvent = new();
+
+        private ControlBindingsCollection? _dataBindings;
+        private BindingContext? _bindingContext;
+
+        [Browsable(false)]
+        [EditorBrowsable(EditorBrowsableState.Advanced)]
+        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
+        [SRCategory(nameof(SR.CatData))]
+        [SRDescription(nameof(SR.ToolStripItemBindingContextDescr))]
+        public BindingContext BindingContext
+        {
+            get => _bindingContext ??= new BindingContext();
+            set
+            {
+                if (!Equals(_bindingContext, value))
+                {
+                    _bindingContext = value;
+                    OnBindingContextChanged(EventArgs.Empty);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Occurs when the binding context has changed
+        /// </summary>
+        [SRCategory(nameof(SR.CatData))]
+        [SRDescription(nameof(SR.ToolStripItemOnBindingContextChangedDescr))]
+        [EditorBrowsable(EditorBrowsableState.Advanced)]
+        public event EventHandler BindingContextChanged
+        {
+            add => Events.AddHandler(s_bindingContextChangedEvent, value);
+            remove => Events.RemoveHandler(s_bindingContextChangedEvent, value);
+        }
+
+        /// <summary>
+        ///  Raises the <see cref="BindableComponent.BindingContextChanged"/> event.
+        ///  Inheriting classes should override this method to handle this event.
+        ///  Call base.OnBindingContextChanged to send this event to any registered event listeners.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Advanced)]
+        protected virtual void OnBindingContextChanged(EventArgs e)
+        {
+            if (_bindingContext is not null)
+            {
+                for (int i = 0; i < DataBindings.Count; i++)
+                {
+                    BindingContext.UpdateBinding(BindingContext, DataBindings[i]);
+                }
+            }
```

Comment:
Mixing backing fields (i.e., `_bindingContext`) and properties (i.e., `BindingContext`) often makes it harder to reason about the state. Can we use one or the other?
Also using `DataBindings` may force creation of the backing field, so should this be as follows?
```suggestion
            if (_bindingContext is not null && _dataBindings is not null)
            {
                for (int i = 0; i < _dataBindings.Count; i++)
                {
                    _bindingContext.UpdateBinding(_bindingContext, _dataBindings[i]);
                }
            }
```

Btw, this looks very strange: `BindingContext.UpdateBinding(BindingContext, ...)`

Reply:
In this case `BindingContext` is the type name - we're calling the _static_ method 'UpdateBinding'. It may look a bit weird, but it should be fine like this.